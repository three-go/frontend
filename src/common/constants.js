const iconNames = {
  heart: "heart",
  brokenHeart: "heart-broken-outline",
  leftArrow: "arrow-left-thick",
  flag: "flag",
};

const iconSizes = {
  headerChance: 28,
  headerLeftArrow: 40,
  headerFlag: 32,
};

const colors = {
  black: "#000000",
  dark: "#212529",
  blueGray: "#2F455C",
  ligthGray: "#5c6977",
  gray: "#96A1A8",
  ivory: "#FCF8F6",
  white: "#ffffff",
  red: "#f45a5a",
  darkRed: "#c92a2a",
  green: "#21D0B2",
  tealGreen: "#00BBD1",
  turquoise: "#00E0EA",
  lightBlue: "#5fceec",
};

const gameStatus = {
  ready: "ready",
  start: "start",
  input: "input",
  end: "end",
};

const navigations = {
  main: "Main",
};

const map = {
  width: 300,
  height: 450,
};

const facePosition = {
  top: 60,
  bottom: 390,
};

const faceYawAngleRange = {
  front: {
    low: -10,
    high: 10,
  },
  left: -36,
  right: 36,
};

export {
  iconNames,
  iconSizes,
  navigations,
  map,
  colors,
  facePosition,
  faceYawAngleRange,
};
