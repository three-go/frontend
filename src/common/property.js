import uuid from "react-native-uuid";

const gameNames = {
  game1: "GAME 1",
  game2: "GAME 2",
};

const gameColors = {
  game1: "#00BBD1",
  game2: "#00E0EA",
};

const game1Guide = [
  {
    title: "게임 설명",
    titleContent:
      "• 시작지점부터 종료지점까지 장애물을 피해 미로를 탈출하는 게임입니다.\n• 총 3개의 stage로 구성되어 있습니다.",
    children: [
      {
        id: uuid.v4(),
        title: "Map",
        content:
          "• 일정시간동안 맵을 기억하세요.\n• 탈출하는 최단거리를 기억하세요",
        image: require("../../public/assets/images/game1/map.jpg"),
      },
      {
        id: uuid.v4(),
        title: "Face",
        content: "• 얼굴로 캐릭터가 움직이는 방향을 기록하세요.",
        image: require("../../public/assets/images/game1/camera.png"),
        icon: "face-recognition",
      },
      {
        id: uuid.v4(),
        title: "Play",
        content: "• 기록된 방향에 따라 캐릭터가 움직입니다.",
        image: require("../../public/assets/images/game1/play.jpg"),
      },
      {
        id: uuid.v4(),
        title: "Next Stage",
        content: "• 캐릭터가 종료지점에 도착하면 다음 스테이지로 이동합니다.",
        image: require("../../public/assets/images/game1/next.jpg"),
      },
      {
        id: uuid.v4(),
        title: "Fail",
        content:
          "• 캐릭터가 종료지점에 도착하지 않으면 목숨 하나가 소진된다.\n• 다시시도 또는 메인화면으로 돌아갈수 있다.",
        image: require("../../public/assets/images/game1/fail.jpg"),
      },
      {
        id: uuid.v4(),
        title: "End",
        content: "• 목숨을 모두 소진하면 게임이 종료된다.",
        image: require("../../public/assets/images/game1/end.jpg"),
      },
      {
        id: uuid.v4(),
        title: "Success",
        content: "• 마지막 stage까지 성공하면 내 점수를 입력할 수 있습니다.",
        image: require("../../public/assets/images/game1/success.jpg"),
      },
      {
        id: uuid.v4(),
        title: "Score",
        content: "• 지금까지 플래이한 점수를 볼 수 있습니다.",
        image: require("../../public/assets/images/game1/score.jpg"),
      },
    ],
  },
  {
    title: "플레이 방법",
    titleContent: "• 게임 진행 방법에 대하여 알려드리겠습니다",
    children: [
      {
        id: uuid.v4(),
        title: "Face",
        content:
          "• 위쪽: 얼굴의 눈 위치를 카메라 화면에 상단에 위치시키세요\n• 아래쪽: 얼굴의 눈 위치를 카메라 화면에 하단에 위치시키세요\n• 왼쪽: 얼굴의 고개를 왼쪽으로 돌리세요.\n• 오른쪽: 얼굴의 고개를 오른쪽으로 돌리세요.",
        image: require("../../public/assets/images/game1/camera.png"),
        icon: "face-recognition",
      },
      {
        id: uuid.v4(),
        title: "Play",
        content: "• 기록된 방향에따라 캐릭터가 자동으로 움직입니다.",
        image: require("../../public/assets/images/game1/play.jpg"),
      },
    ],
  },
];

const gameDescription = {
  game1: game1Guide,
  game2: [],
};

const gameKey = ["game1", "game2"];

const game = {
  keys: gameKey,
  names: gameNames,
  colors: gameColors,
  description: gameDescription,
};

const colors = {
  ivory: "#FCF8F6",
};

export { game, colors };
