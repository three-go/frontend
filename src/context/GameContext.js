import React, { createContext, useState } from "react";

const GameContext = createContext();

export const GameContextProvider = ({ children }) => {
  const [currentGameKey, setCurrentGameKey] = useState("");

  return (
    <GameContext.Provider value={{ currentGameKey, setCurrentGameKey }}>
      {children}
    </GameContext.Provider>
  );
};

export default GameContext;
