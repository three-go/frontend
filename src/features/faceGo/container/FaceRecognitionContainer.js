import React, { useState, useEffect } from "react";

import PropTypes from "prop-types";
import { Platform } from "react-native";
import uuid from "react-native-uuid";

import {
  facePosition,
  faceYawAngleRange,
  time,
} from "../../../common/constants";
import FaceRecognition from "../presenter/FaceRecognition";

const FaceRecognitionContainer = ({
  selectedDirection,
  onSelectedDirection,
  cameraPermissionStatus,
  setCameraPermissionStatus,
}) => {
  const [isFaceCenter, setIsFaceCenter] = useState(false);
  const [isFaceFront, setIsFaceFront] = useState(false);
  const [isDirectionPreviewTextShow, setIsDirectionPreviewTextShow] =
    useState(true);

  useEffect(() => {
    const timeoutId = setTimeout(() => {
      setIsDirectionPreviewTextShow(false);
    }, time.faceRecognitionInterval);

    return () => {
      clearTimeout(timeoutId);
    };
  }, [selectedDirection]);

  const handlerFacePosition = ({ faces }) => {
    const focusedFace = faces[0];

    if (!focusedFace) {
      return;
    }

    if (
      !isFaceCenter &&
      focusedFace.leftEyePosition.y > facePosition.top &&
      focusedFace.leftEyePosition.y < facePosition.bottom
    ) {
      setIsFaceCenter(true);
    }

    if (
      !isFaceFront &&
      focusedFace.yawAngle > faceYawAngleRange.front.low &&
      focusedFace.yawAngle < faceYawAngleRange.front.high
    ) {
      setIsFaceFront(true);
    }

    if (isFaceCenter && isFaceFront) {
      if (focusedFace.leftEyePosition.y < facePosition.top) {
        setIsFaceCenter(false);
        setIsDirectionPreviewTextShow(true);
        onSelectedDirection({ id: uuid.v4(), direction: "up" });
      }

      if (focusedFace.leftEyePosition.y > facePosition.bottom) {
        setIsFaceCenter(false);
        setIsDirectionPreviewTextShow(true);
        onSelectedDirection({ id: uuid.v4(), direction: "down" });
      }

      if (focusedFace.yawAngle < faceYawAngleRange.left) {
        setIsFaceFront(false);
        setIsDirectionPreviewTextShow(true);
        Platform.OS === "ios"
          ? onSelectedDirection({ id: uuid.v4(), direction: "left" })
          : onSelectedDirection({ id: uuid.v4(), direction: "right" });
      }

      if (focusedFace.yawAngle > faceYawAngleRange.right) {
        setIsFaceFront(false);
        setIsDirectionPreviewTextShow(true);
        Platform.OS === "ios"
          ? onSelectedDirection({ id: uuid.v4(), direction: "right" })
          : onSelectedDirection({ id: uuid.v4(), direction: "left" });
      }
    }
  };

  return (
    <FaceRecognition
      selectedDirection={selectedDirection}
      handlerFacePosition={handlerFacePosition}
      cameraPermissionStatus={cameraPermissionStatus}
      setCameraPermissionStatus={setCameraPermissionStatus}
      isDirectionPreviewTextShow={isDirectionPreviewTextShow}
    />
  );
};

export default FaceRecognitionContainer;

FaceRecognitionContainer.propTypes = {
  selectedDirection: PropTypes.shape({
    id: PropTypes.string,
    direction: PropTypes.string,
  }),
  onSelectedDirection: PropTypes.func.isRequired,
  cameraPermissionStatus: PropTypes.string.isRequired,
  setCameraPermissionStatus: PropTypes.func.isRequired,
};

FaceRecognitionContainer.defaultProps = {
  selectedDirection: {},
};
