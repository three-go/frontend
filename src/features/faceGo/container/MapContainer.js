import React, { useEffect } from "react";

import useCharacter from "../../../hooks/useCharacter";
import { map, game } from "../../../common";
import { Map } from "../presenter";

const MapContainer = ({
  status,
  gameMap,
  stage,
  directions,
  setResult,
  score,
  setScore,
  setChance,
}) => {
  const characterInfo = useCharacter(gameMap, score, setScore);

  const arrInfo = {
    columnCount: gameMap && gameMap[0].length,
    rowCount: gameMap && gameMap.length,
  };

  const borderWidth = {
    horizontal: 20 + 2 * arrInfo.columnCount,
    vertical: 15 + 2 * arrInfo.rowCount,
  };

  const boxStyle = {
    boxWidth: (map.width - borderWidth.horizontal) / arrInfo.columnCount,
    boxHeigth: (map.height - borderWidth.vertical) / arrInfo.rowCount,
  };

  const handleCheckStage = (n) => {
    if (n < 3) {
      setResult(game.result.win);
    } else if (stage === 3) {
      setResult(game.result.end);
    }
  };

  useEffect(() => {
    if (status === game.status.play && directions.length === 0) {
      const { x, y } = characterInfo.position;

      if (x === arrInfo.columnCount - 1 && y === arrInfo.rowCount - 1) {
        handleCheckStage(stage);
      } else {
        setChance((prev) => prev - 1);
        setResult(game.result.lose);
      }
    }

    if (!directions[0]) {
      return;
    }

    switch (status === game.status.play && directions[0].direction) {
      case "left":
        characterInfo.moveLeft();
        break;
      case "right":
        characterInfo.moveRight();
        break;
      case "up":
        characterInfo.moveUp();
        break;
      case "down":
        characterInfo.moveDown();
        break;
      default:
        "left";
    }
  }, [directions]);

  return (
    <Map
      status={status}
      gameMap={gameMap}
      characterInfo={characterInfo}
      arrInfo={arrInfo}
      boxStyle={boxStyle}
      directions={directions}
      setScore={setScore}
    />
  );
};

export default MapContainer;
