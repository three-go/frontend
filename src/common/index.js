import {
  iconNames,
  iconSizes,
  navigations,
  map,
  colors,
  facePosition,
  faceYawAngleRange,
} from "./constants";
import { game } from "./property";

export {
  iconNames,
  iconSizes,
  navigations,
  map,
  colors,
  facePosition,
  faceYawAngleRange,
  game,
};
