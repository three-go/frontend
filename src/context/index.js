import GameContext, { GameContextProvider } from "./GameContext";

export { GameContext, GameContextProvider };
