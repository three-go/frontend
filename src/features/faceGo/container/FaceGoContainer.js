import React, { useState, useEffect, useContext, useMemo } from "react";

import { FaceGo } from "..";
import { game, time } from "../../../common";
import { GameContext } from "../../../context";
import { createMap } from "../../../utils";

const FaceGoContainer = () => {
  const { currentGameKey } = useContext(GameContext);

  const [status, setStatus] = useState(game.status.none);
  const [result, setResult] = useState(game.result.none);

  const [cameraPermissionStatus, setCameraPermissionStatus] = useState("");
  const [selectedDirection, setSelectedDirection] = useState({});
  const [directions, setDirections] = useState([]);

  const [score, setScore] = useState(500);
  const [stage, setStage] = useState(1);
  const [chance, setChance] = useState(3);

  const gameMap = useMemo(() => {
    return createMap(stage);
  }, [stage]);

  const handleReset = () => {
    setStatus("none");
    setSelectedDirection((prev) => {});
  };

  const handleNextStage = () => {
    setStage((prev) => prev + 1);
    setResult(game.result.none);
    handleReset();
  };

  const handleRetryStage = () => {
    setResult(game.result.none);
    handleReset();
  };

  const handleSetStatusOpen = () => {
    setStatus(game.status.open);
  };

  useEffect(() => {
    if (!selectedDirection?.direction) {
      return;
    }

    setDirections((prevState) => [...prevState, selectedDirection]);
  }, [selectedDirection]);

  useEffect(() => {
    let timeoutId;

    if (status === game.status.play) {
      if (directions.length < 1) {
        return;
      }

      timeoutId = setTimeout(() => {
        const copy = directions.slice();
        copy.shift();
        setDirections(copy);
      }, time.playInterval);
    }

    return () => {
      clearTimeout(timeoutId);
    };
  }, [status, directions.length]);

  useEffect(() => {
    if (status !== game.status.play) {
      game.sounds.timer.setNumberOfLoops(-1);
      game.sounds.timer.play();
    }

    return () => {
      if (status !== game.status.play) {
        game.sounds.timeout.play();
        game.sounds.timer.stop();
      }
    };
  }, [status]);

  return (
    <FaceGo
      gameInfo={{
        stage,
        status,
        setStatus,
        result,
        setResult,
        currentGameKey,
        gameMap,
        handleNextStage,
        handleRetryStage,
        handleSetStatusOpen,
      }}
      userInfo={{
        score,
        setScore,
        chance,
        setChance,
        directions,
        setDirections,
      }}
      cameraInfo={{
        selectedDirection,
        setSelectedDirection,
        cameraPermissionStatus,
        setCameraPermissionStatus,
      }}
    />
  );
};

export default FaceGoContainer;
