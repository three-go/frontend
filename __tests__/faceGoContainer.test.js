import React from "react";

import { render } from "@testing-library/react-native";
import "@testing-library/jest-native";
import "react-native-gesture-handler/jestSetup";

import { camera } from "../src/common/constants";
import { game } from "../src/common/property";
import { GameContextProvider } from "../src/context/GameContext";
import FaceGoContainer from "../src/features/faceGo/container/FaceGoContainer";
import FaceRecognitionContainer from "../src/features/faceGo/container/FaceRecognitionContainer";
import MapContainer from "../src/features/faceGo/container/MapContainer";
import { createMap } from "../src/utils/map";

const mockNavigate = jest.fn();
const mockDisPatch = jest.fn();

jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockNavigate,
      dispatch: mockDisPatch,
    }),
  };
});

jest.mock("react-native-reanimated", () => {
  const Reanimated = require("react-native-reanimated/mock");
  Reanimated.default.call = () => {};
  return Reanimated;
});

jest.mock("react-native/Libraries/Animated/NativeAnimatedHelper");

jest.mock("react-native/Libraries/EventEmitter/NativeEventEmitter.js", () =>
  require("react-native/Libraries/EventEmitter/__mocks__/NativeEventEmitter.js")
);

jest.mock("react-native-sound", () => {
  const Sound = () => {
    const obj = {};
    obj.sound = "music";
    obj.setNumberOfLoops = jest.fn();
    obj.play = jest.fn();
    return obj;
  };

  return Sound;
});

jest.mock("@react-native-community/async-storage", (key, value) => {
  const storage = {};
  storage[key] = value;
  return storage;
});

describe("Container Test", () => {
  it("FaceGoContainer Test", () => {
    const faceGoContainer = render(
      <GameContextProvider>
        <FaceGoContainer />
      </GameContextProvider>
    );
    const faceGoContainerJson = faceGoContainer.toJSON();

    expect(faceGoContainer.getByTestId("layout")).toBeTruthy();
    expect(faceGoContainerJson).toBeTruthy();
    expect(faceGoContainerJson).toMatchSnapshot();
  });

  it("FaceRecognitionContainer Test", () => {
    let selectedDirection = {
      id: null,
      direction: null,
    };

    const mockOnSelectedDirection = jest.fn();
    const mockSetCameraPermissionStatus = jest.fn();

    mockOnSelectedDirection.mockReturnValue({
      id: "threeGo333",
      direction: "left",
    });
    mockSetCameraPermissionStatus.mockReturnValue(camera.permissionReady);

    const props = {
      selectedDirection,
      onSelectedDirection: () => {
        selectedDirection = mockOnSelectedDirection();
      },
      cameraPermissionStatus: camera.permissionNotAuthorized,
      setCameraPermissionStatus: mockSetCameraPermissionStatus,
    };

    const faceRecognitionContainer = render(
      <FaceRecognitionContainer {...props} />
    );
    const faceRecognitionContainerJson = faceRecognitionContainer.toJSON();

    expect(
      faceRecognitionContainer.getByTestId("recognition")
    ).toHaveTextContent(
      /권한 승인이 필요 합니다. 해당 서비스를 사용하기 위해서는 카메라 권한 승인이 필요합니다. 해당 권한을 승인하지 않으면 서비스 이용이 제한됩니다./
    );
    expect(faceRecognitionContainerJson).toBeTruthy();
    expect(faceRecognitionContainerJson).toMatchSnapshot();
  });

  it("MapContainer Test", () => {
    global.__reanimatedWorkletInit = () => {};

    const props = {
      stage: 1,
      status: game.status.open,
      setResult: jest.fn(),
      gameMap: createMap(1),
      score: 0,
      setScore: jest.fn(),
      directions: [
        { id: "id1", direction: "right" },
        { id: "id2", direction: "right" },
        { id: "id3", direction: "down" },
        { id: "id4", direction: "down" },
        { id: "id5", direction: "down" },
      ],
      setChance: jest.fn(),
    };

    const mapContainer = render(<MapContainer {...props} />);
    const mapContainerJson = mapContainer.toJSON();

    mapContainer.debug();

    expect(mapContainer.getByTestId("map")).toHaveTextContent("Start");
    expect(mapContainer.getByTestId("map")).toHaveTextContent("End");
    expect(mapContainerJson).toBeTruthy();
    expect(mapContainerJson).toMatchSnapshot();
  });
});
