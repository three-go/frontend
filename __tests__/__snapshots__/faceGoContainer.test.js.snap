// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container Test FaceGoContainer Test 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "top",
    ]
  }
  style={
    Object {
      "backgroundColor": "#212529",
      "flex": 1,
    }
  }
  testID="layout"
>
  <View
    style={
      Object {
        "height": "10%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#212529",
          "color": "#FCF8F6",
          "flexDirection": "row",
          "height": "100%",
          "justifyContent": "space-evenly",
          "paddingLeft": 8,
          "paddingRight": 8,
          "width": "100%",
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "position": "relative",
      }
    }
  />
</RNCSafeAreaView>
`;

exports[`Container Test FaceRecognitionContainer Test 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "height": "100%",
      "justifyContent": "center",
      "width": "100%",
    }
  }
  testID="recognition"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <ActivityIndicator
      size="small"
    />
  </View>
  <Image
    source={
      Object {
        "testUri": "../../../public/assets/images/faceGo/faceShape.png",
      }
    }
    style={
      Object {
        "height": "100%",
        "opacity": 0.3,
        "position": "absolute",
        "width": "100%",
      }
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "borderRadius": 20,
        "flex": 1,
        "height": "50%",
        "justifyContent": "center",
        "padding": "5%",
        "width": "80%",
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 20,
          "fontWeight": "900",
        }
      }
    >
      권한 승인이 필요 합니다.
      

    </Text>
    <Text>
      해당 서비스를 사용하기 위해서는 카메라 권한 승인이 필요합니다. 해당 권한을 승인하지 않으면 서비스 이용이 제한됩니다.
      

    </Text>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#f45a5a",
          "borderRadius": 20,
          "height": 40,
          "justifyContent": "center",
          "opacity": 1,
          "width": 100,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FCF8F6",
            "fontSize": 15,
            "fontWeight": "bold",
            "padding": 8,
          }
        }
      >
        설정으로 가기
      </Text>
    </View>
  </View>
</View>
`;

exports[`Container Test MapContainer Test 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#FCF8F6",
        "borderColor": "#FCF8F6",
        "borderRadius": 10,
        "borderWidth": 5,
        "flexWrap": "wrap",
        "justifyContent": "center",
      },
      [Function],
      [Function],
    ]
  }
  testID="map"
>
  <RCTScrollView
    data={
      Array [
        1,
        1,
        0,
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEnabled={false}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#21D0B2",
                "borderColor": "#ffffff",
                "borderRadius": 10,
                "borderWidth": 2,
                "height": 106.75,
                "justifyContent": "center",
                "width": 91.33333333333333,
              }
            }
          />
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#5fceec",
                "borderColor": "#ffffff",
                "borderRadius": 10,
                "borderWidth": 2,
                "height": 106.75,
                "justifyContent": "center",
                "width": 91.33333333333333,
              }
            }
          />
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#2F455C",
                "borderColor": "#ffffff",
                "borderRadius": 10,
                "borderWidth": 2,
                "height": 106.75,
                "justifyContent": "center",
                "width": 91.33333333333333,
              }
            }
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#5c6977",
                    "fontSize": 91.33333333333333,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              󰅖
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <RCTScrollView
    data={
      Array [
        0,
        1,
        0,
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEnabled={false}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#2F455C",
                "borderColor": "#ffffff",
                "borderRadius": 10,
                "borderWidth": 2,
                "height": 106.75,
                "justifyContent": "center",
                "width": 91.33333333333333,
              }
            }
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#5c6977",
                    "fontSize": 91.33333333333333,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              󰅖
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#5fceec",
                "borderColor": "#ffffff",
                "borderRadius": 10,
                "borderWidth": 2,
                "height": 106.75,
                "justifyContent": "center",
                "width": 91.33333333333333,
              }
            }
          />
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#2F455C",
                "borderColor": "#ffffff",
                "borderRadius": 10,
                "borderWidth": 2,
                "height": 106.75,
                "justifyContent": "center",
                "width": 91.33333333333333,
              }
            }
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#5c6977",
                    "fontSize": 91.33333333333333,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              󰅖
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <RCTScrollView
    data={
      Array [
        1,
        1,
        1,
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEnabled={false}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#5fceec",
                "borderColor": "#ffffff",
                "borderRadius": 10,
                "borderWidth": 2,
                "height": 106.75,
                "justifyContent": "center",
                "width": 91.33333333333333,
              }
            }
          />
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#5fceec",
                "borderColor": "#ffffff",
                "borderRadius": 10,
                "borderWidth": 2,
                "height": 106.75,
                "justifyContent": "center",
                "width": 91.33333333333333,
              }
            }
          />
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#5fceec",
                "borderColor": "#ffffff",
                "borderRadius": 10,
                "borderWidth": 2,
                "height": 106.75,
                "justifyContent": "center",
                "width": 91.33333333333333,
              }
            }
          />
        </View>
      </View>
    </View>
  </RCTScrollView>
  <RCTScrollView
    data={
      Array [
        0,
        1,
        2,
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEnabled={false}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#2F455C",
                "borderColor": "#ffffff",
                "borderRadius": 10,
                "borderWidth": 2,
                "height": 106.75,
                "justifyContent": "center",
                "width": 91.33333333333333,
              }
            }
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#5c6977",
                    "fontSize": 91.33333333333333,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              󰅖
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#5fceec",
                "borderColor": "#ffffff",
                "borderRadius": 10,
                "borderWidth": 2,
                "height": 106.75,
                "justifyContent": "center",
                "width": 91.33333333333333,
              }
            }
          />
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#21D0B2",
                "borderColor": "#ffffff",
                "borderRadius": 10,
                "borderWidth": 2,
                "height": 106.75,
                "justifyContent": "center",
                "width": 91.33333333333333,
              }
            }
          />
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "left": 10,
        "position": "absolute",
        "top": 5,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FCF8F6",
          "fontSize": 14,
        }
      }
    >
      Start
    </Text>
  </View>
  <View
    style={
      Object {
        "bottom": 5,
        "position": "absolute",
        "right": 15,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FCF8F6",
          "fontSize": 14,
        }
      }
    >
      End
    </Text>
  </View>
</View>
`;
